### PR TITLE
Expo-ota unimodule for bare applications

### PR DESCRIPTION
**Pull request by mczernek**
# Why

Created from scratch new unimodule for handling over the air updates.

# Test Plan

Test application available in this PR: https://github.com/expo/expo/pull/5999

### CODE REVIEW

**Review by esamelson — CHANGES_REQUESTED** [review]
Thanks for all your work on this! This is going to take a few passes and some time for me to digest. I left a few small surface-level comments, but here are a few larger-scale thoughts after a first pass:
- Maybe you've already considered this and I'm just missing it, but a critical piece of functionality here is being able to load the manifest and bundle embedded in the IPA/APK. This is important in a number of scenarios, including (but not limited to) the first time the app is launched (with or without an internet connection) and when the IPA/APK is updated. There are a number of edge cases here that must be handled correctly and afaik the only way to do this is to compare the `commitTime`s and `sdkVersion`s on the cached manifest and embedded manifest every time the app is launched. It's imperative we **never** load an older JS bundle after having loaded a newer one because, for example, there is no guarantee developers will persist backwards-compatible data.
Edge case 1: dev publishes bundle A with build 1, then publishes bundle B as an OTA update. Then creates build 2 with bundle C which is newer (and not backwards compatible). A user has bundle B cached as an OTA update but then updates their binary to build 2 through the App/Play Store. As soon as they launch build 2 we need to use the embedded bundle C immediately and ignore the cached bundle B.
Edge case 2: dev publishes bundle A with build 1, then publishes bundle B along with build 2. Afterwards, they publish bundle C as an OTA update. A user has build 1 and gets the OTA update to bundle C. Later, they update their binary to build 2. Since cached bundle C is newer than embedded bundle B, we need to load C if it's compatible with build 2 (i.e. if the SDK version matches), but if not, then we are forced to load B.
Currently, when you run `expo publish` in an ExpoKit project, expo-cli copies the manifest + bundle to a location on disk that allows them to be bundled in the native binaries. We may need to modify this to make it work for bare projects as well.
^ Sorry if that is a bit of information overload, but just wanted to give a sense of why this functionality is critical! And double apologies if you've already thought of all this and I managed to overlook it.
- We need to make the manifest available to the JS layer. Currently we do this through the Constants module, but we can add it to this module if that's easier.
- It looks like this module is not taking into account the `updates` configuration from the manifest/app.json. This is perhaps less critical than the previous two points, but if we enforce an opinionated configuration we need to make sure it's workable for our partners, at least. Is the current implementation equivalent to setting `fallbackToCacheTimeout: 0`?

**Comment by mczernek** [comment]
> Thanks for all your work on this! This is going to take a few passes and some time for me to digest. I left a few small surface-level comments, but here are a few larger-scale thoughts after a first pass:
> 
> * Maybe you've already considered this and I'm just missing it, but a critical piece of functionality here is being able to load the manifest and bundle embedded in the IPA/APK. This is important in a number of scenarios, including (but not limited to) the first time the app is launched (with or without an internet connection) and when the IPA/APK is updated. There are a number of edge cases here that must be handled correctly and afaik the only way to do this is to compare the `commitTime`s and `sdkVersion`s on the cached manifest and embedded manifest every time the app is launched. It's imperative we **never** load an older JS bundle after having loaded a newer one because, for example, there is no guarantee developers will persist backwards-compatible data.
>   Edge case 1: dev publishes bundle A with build 1, then publishes bundle B as an OTA update. Then creates build 2 with bundle C which is newer (and not backwards compatible). A user has bundle B cached as an OTA update but then updates their binary to build 2 through the App/Play Store. As soon as they launch build 2 we need to use the embedded bundle C immediately and ignore the cached bundle B.
>   Edge case 2: dev publishes bundle A with build 1, then publishes bundle B along with build 2. Afterwards, they publish bundle C as an OTA update. A user has build 1 and gets the OTA update to bundle C. Later, they update their binary to build 2. Since cached bundle C is newer than embedded bundle B, we need to load C if it's compatible with build 2 (i.e. if the SDK version matches), but if not, then we are forced to load B.
>   Currently, when you run `expo publish` in an ExpoKit project, expo-cli copies the manifest + bundle to a location on disk that allows them to be bundled in the native binaries. We may need to modify this to make it work for bare projects as well.
>   ^ Sorry if that is a bit of information overload, but just wanted to give a sense of why this functionality is critical! And double apologies if you've already thought of all this and I managed to overlook it.
> * We need to make the manifest available to the JS layer. Currently we do this through the Constants module, but we can add it to this module if that's easier.
> * It looks like this module is not taking into account the `updates` configuration from the manifest/app.json. This is perhaps less critical than the previous two points, but if we enforce an opinionated configuration we need to make sure it's workable for our partners, at least. Is the current implementation equivalent to setting `fallbackToCacheTimeout: 0`?

**Comment by mczernek** [comment]
> Thanks for all your work on this! This is going to take a few passes and some time for me to digest. I left a few small surface-level comments, but here are a few larger-scale thoughts after a first pass:
> 
> * Maybe you've already considered this and I'm just missing it, but a critical piece of functionality here is being able to load the manifest and bundle embedded in the IPA/APK. This is important in a number of scenarios, including (but not limited to) the first time the app is launched (with or without an internet connection) and when the IPA/APK is updated. There are a number of edge cases here that must be handled correctly and afaik the only way to do this is to compare the `commitTime`s and `sdkVersion`s on the cached manifest and embedded manifest every time the app is launched. It's imperative we **never** load an older JS bundle after having loaded a newer one because, for example, there is no guarantee developers will persist backwards-compatible data.
>   Edge case 1: dev publishes bundle A with build 1, then publishes bundle B as an OTA update. Then creates build 2 with bundle C which is newer (and not backwards compatible). A user has bundle B cached as an OTA update but then updates their binary to build 2 through the App/Play Store. As soon as they launch build 2 we need to use the embedded bundle C immediately and ignore the cached bundle B.
>   Edge case 2: dev publishes bundle A with build 1, then publishes bundle B along with build 2. Afterwards, they publish bundle C as an OTA update. A user has build 1 and gets the OTA update to bundle C. Later, they update their binary to build 2. Since cached bundle C is newer than embedded bundle B, we need to load C if it's compatible with build 2 (i.e. if the SDK version matches), but if not, then we are forced to load B.
>   Currently, when you run `expo publish` in an ExpoKit project, expo-cli copies the manifest + bundle to a location on disk that allows them to be bundled in the native binaries. We may need to modify this to make it work for bare projects as well.
>   ^ Sorry if that is a bit of information overload, but just wanted to give a sense of why this functionality is critical! And double apologies if you've already thought of all this and I managed to overlook it.
> * We need to make the manifest available to the JS layer. Currently we do this through the Constants module, but we can add it to this module if that's easier.
> * It looks like this module is not taking into account the `updates` configuration from the manifest/app.json. This is perhaps less critical than the previous two points, but if we enforce an opinionated configuration we need to make sure it's workable for our partners, at least. Is the current implementation equivalent to setting `fallbackToCacheTimeout: 0`?

Thanks a lot for this comment! Honestly speaking, I've considered some but not all concerns you're mentioning. Let me describe my assumptions, maybe this will bring some clarity into the picture.

I assumed, that we deliver two ways (which might cooperate) for users to handle updating of their apps.
1. Automatic. They just do initial config on native site (which should be really easy) and then, whenever they publish their javascript app, it is being updated on users devices.
This workflow requires answering some questions.
- there is no concept of expo sdk version in bare apps. However, we might use that field (or add another one) to let users decide which versions are compatible with which native implementations. We cannot support it as seemlessly as we do in ExpoKit, since we don't really know whether changes in native code appeared and if they are braking. Expo Developers are in control here.
- what about initial version and manifest? manifest from app.json differs from this delivered by updates API. Can we treat them equally?
2. Manual. We provide users with set of methods for users, which lets them check for available manifest, as well as some configuration on native side. Users are in full control of when update is being performed if at all. In this scenario, we rely on developers shaping their code in such manner, that updates are performed carefully. We can only fit them with proper tools to do so.

I believe both of those variants might benefit from last two points and will work on that. I already made some naive implementation of bundling manifest into native code and upload only when needed.

It might be also relevant to discuss current API of the module. It is widely based on what's available now, but this might be subject of discussion and change.

**Comment by esamelson** [comment]
Awesome, makes sense. As much as possible, we'd like the API of this module to stay the same as the current OTA updates capabilities in ExpoKit -- and I mean this in the broadest sense, not only the actual methods exposed to JS through the module, but also the different configuration options (automatic/manual like you mentioned, probably through app.json) and being able to publish to expo's servers using expo publish. (The new updates service we're planning/working on in parallel will comprise a new backend service as well as a redesigned client implementation, but we need this module in the meantime to allow people--including important partners--to move away from ExpoKit and to the bare workflow while maintaining as much compatibility as possible with our current OTA updates client API and backend.)

=> a couple of things that I think result from this that address some of your points:
- I think we need to keep using SDK version for now, since it's pretty tied into how our current backend works for serving published OTA updates. I know it's not a perfect fit since people can make breaking changes to their native code in the bare workflow, but I think for now it's ok to leave it up to developers to deal with this properly.
- Similarly, I think it's fine to require a similar app.json to allow expo-cli to publish to Expo's servers. In the current implementation, expo-cli downloads the published version of the manifest to stick into the app binary, rather than just copying app.json.

I agree that in the future we'll need to give more agency/flexibility to developers, especially with versioning, but I think that this is the best path forward for now to stay compatible with our current system and minimize churn for developers. Let me know what you think!

**Review by esamelson — COMMENTED** [review]
still will need to do at least one more pass to really grep everything, but getting a better picture 😄 things are looking good, thanks for making all of these improvements! Here are a couple of thoughts in addition to the comments.

- I think we need to be comparing `commitTime` rather than `version` number or `revisionId `-- this is going to be more accurate and have a more consistent format.
- there are a number of formatting issues in the obj-c files. could you make sure you have XCode set to a tab size of 2?

File: packages/expo-ota/android/src/main/java/expo/modules/ota/OtaUpdater.kt
Comment (esamelson) [inline packages/expo-ota/android/src/main/java/expo/modules/ota/OtaUpdater.kt]: is it possible that we want to set `persistence.enqueuedReorderAtNextBoot = false` here? Seems like it might be safe since if we get to this point we know which bundle we are going to run. Will defer to you on this though.

File: packages/expo-ota/android/src/main/java/expo/modules/ota/OtaModule.kt
Comment (esamelson) [inline packages/expo-ota/android/src/main/java/expo/modules/ota/OtaModule.kt]: looks like reload may still need to be implemented. Is this doable?

File: packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt
Comment (esamelson) [inline packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt]: One thing that's really important is that the manifest OkHttpClient should never read from a cache. Currently in ExpoKit we use `CacheControl.FORCE_NETWORK` to enforce this. Otherwise we might end up missing an update, since the manifest url, in theory, never changes.

On the other hand, for the validator we have pretty specific logic to use a cache for the public key and then only retry with network if the validation fails (in case we rotated the keys). Would be nice to replicate that logic if possible. https://github.com/expo/expo/blob/380493569893c76fe905c3295bcbfb3cc3bb3865/android/expoview/src/main/java/host/exp/exponent/kernel/Crypto.java#L52-L56

**Review by esamelson — COMMENTED** [review]
Sorry this review is taking so long -- I'm getting a clearer picture of things now, it's just taking multiple passes 😞 Generally things are looking pretty good! Left some more comments on the Android side. Will take another closer look at iOS next (though the comments I left also apply conceptually to iOS).

File: packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt
Comment (mczernek) [reply packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt]: Sure, sounds like plan :)

File: packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt
Comment (mczernek) [reply packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt]: You are right, my bad. Will change.

File: packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt
Comment (mczernek) [reply packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt]: I believe that for the cache to be used, server would need to response with `304` response code. It might change once we set `Cache-Control` to 'Force-Cache`, but as long as there is no forcing, it should work automatically right. And as far as I was testing it, this is the case. Even for very frequent updates (like within 10 seconds), none of updates were missing.

When it comes to caching public key, I believe that we can reimplement that behavior.

File: packages/expo-ota/android/src/main/java/expo/modules/ota/OtaModule.kt
Comment (mczernek) [reply packages/expo-ota/android/src/main/java/expo/modules/ota/OtaModule.kt]: This is tricky.

While on Android that is easily doable and there are libraries for that (e. g. https://github.com/JakeWharton/ProcessPhoenix which I tested and it worked), on iOS we need to do it within AppDelegate.

My decision was, that we will keep it out of scope of this library, and will let users to decide which solution to use to reload their apps. Maybe we could provide them another module `expo-reload`?

This is not final decision, surely, and I will gladly discuss that. Just wanted to let you know what my view on that matter is.

Comment (esamelson) [reply packages/expo-ota/android/src/main/java/expo/modules/ota/OtaModule.kt]: Hmmm, ok. Let's leave this as a TODO for now so it doesn't block this PR from landing.

File: package.json
Comment (brentvatne) [inline package.json]: does this need to be a root level dep?

**Comment by ide** [comment]
I didn't read over the code but before publishing we need to rename this to expo-updates (v0.0.1 or v0.1.0 -- not 1.0.0).

File: packages/expo-ota/ios/EXOta/EXEmbeddedManifestAndBundle.m
Comment (brentvatne) [inline packages/expo-ota/ios/EXOta/EXEmbeddedManifestAndBundle.m]: this causes the app to crash immediately if we don't have a `shell-app-manifest.json` embedded in the ios app. this isn't explained in readme and doesn't seem ideal - might make sense to just have app.json there and give a good error message if the user hasn't included it in their project

**Comment by brentvatne** [comment]
i published 0.0.1-alpha.0 of expo-ota from this branch and tried to get it running in a new bare project on ios: https://github.com/brentvatne/ota-test

i was unable to get the app running with expo-ota installed, it crashes at startup in debug mode. it seems that we're expecting `shell-app-manifest.json` to be present even in debug. could you get the module running on that repo and open a pr? in exchange, i can update the readme here accordingly :)

File: packages/expo-ota/README.md
Comment (brentvatne) [inline packages/expo-ota/README.md]: we need to keep the meaning of `sdkVersion` the same as in the managed workflow, because if we overload it to mean the binary version then people can't [use the expo client with their bare app](https://docs.expo.io/versions/v35.0.0/bare/using-expo-client/).

Comment (mczernek) [reply packages/expo-ota/README.md]: Honestly, I was unaware that we support such scenario. But if we do, we need to tweak our logic accordingly.

I will update README and app.json schema with new field, `nativeVersion`, which will be required in bare apps (I do not know how to enforce that, yet).
I will compare downloaded and embedded manifests against both `sdkVersion` and `nativeVersion`, given they're both present.
We should add support for native version header in our updates API, which would work similar to sdkVersion. Let us discuss how we would approach that.

Comment (mczernek) [reply packages/expo-ota/README.md]: After discussion with Stanisław we suggest, that we keep it as it is. W leave current meaning of `sdkVersion` and do not add any new field for keeping track of user's native code version whatsoever.

If users will want to distinguish which apps uses which bundles, they can use release channels for now, and we'll try to document that best we can.

Comment (brentvatne) [reply packages/expo-ota/README.md]: sgtm

File: package.json
Comment (mczernek) [reply package.json]: No, it does not :)

**Comment by slorber** [comment]
Has this been merged/released? :)

**Pull request closed without merging**
